core/src/main/scala/org/broadinstitute/dsde/rawls/entities/EntityService.scala

```diff
@@ -515,101 +521,206 @@ class EntityService(protected val ctx: RawlsRequestContext,
+                                      shardId: String,
+                                      boundary: MigrationBoundary,
+                                      dataAccess: DataAccess,
+                                      batchIdx: Int,
+                                      totalBatches: Int,
+                                      stopwatch: StopWatch,
+                                      parentContext: RawlsRequestContext
+  ): ReadWriteAction[Int] = {
+
+    // instrumentation helper
+    def logAndTrace[T, E <: Effect](spanName: String, logMessage: String)(
+      op: DBIOAction[T, NoStream, E]
+    ): DBIOAction[T, NoStream, E with Effect] =
+      traceDBIOWithParent(spanName, parentContext) { _ =>
+        op
+      }.map { result =>
+        logger.info(
+          s"Quicksilver migration $workspaceId batch ${batchIdx + 1}/$totalBatches: $logMessage (${stopwatch.formatTime()}) ..."
+        )
+        result
+      }
 
-      // return a count of entities updated
-    } yield entityTypeMetadata.map { case (entityType, metadata) => (entityType, metadata.count) }
+    for {
+      // create temp tables
+      _ <- logAndTrace("migrationCreateTempTables", "created migration temp tables") {
+        DBIO.seq(dataAccess.compactEntityQuery.migrationCreateAttributeTempTable,
+                 dataAccess.compactEntityQuery.migrationCreateEntityTempTable
+        )
+      }
+      // normalize attributes to JSON scalars and insert into the temp table
+      _ <- logAndTrace("migrationPopulateAttributeTempTable", "populated attribute temp table") {
+        dataAccess.compactEntityQuery.migrationPopulateAttributeTempTable(workspaceId,
+                                                                          shardId,
+                                                                          boundary.startEntityId,
+                                                                          boundary.endEntityId
+        )
+      }
+      // combine scalars into arrays; join all attributes into a single JSON object per entity; populate entity temp
+      _ <- logAndTrace("migrationPopulateEntityTempTable", "populated entity temp table") {
+        dataAccess.compactEntityQuery.migrationPopulateEntityTempTable
+      }
+
+      // update ENTITY from the contents of the temp table
+      numEntitiesUpdated <- logAndTrace("migrationUpdateEntityTable", "updated ENTITY from temp table") {
+        dataAccess.compactEntityQuery.migrationUpdateEntityTable(workspaceId)
+      }
+      // drop temp tables
+      // this explicitly uses create/drop table instead of truncate to avoid implicit transaction commits:
+      // https://dev.mysql.com/doc/refman/8.0/en/implicit-commit.html
+      _ <- logAndTrace("migrationDropTempTables", "dropped temp tables") {
+        DBIO.seq(dataAccess.compactEntityQuery.migrationDropAttributeTempTable,
+                 dataAccess.compactEntityQuery.migrationDropEntityTempTable
```

Comment:
try `DBIOAction.andFinally`, e.g.
```
(<for comp>).andFinally(DBIO.seq(drops))
```

Reply:
done in [9e706d0](https://github.com/broadinstitute/rawls/pull/3325/commits/9e706d0b21f7b0fc4bb43215674076fbc24a5709)